swarm/src/connection.rs

```diff
@@ -431,17 +457,18 @@ where
                 }
             }
 
-            let new_protocols = gather_supported_protocols(handler);
-            let changes = ProtocolsChange::from_full_sets(supported_protocols, &new_protocols);
-
-            if !changes.is_empty() {
-                for change in changes {
-                    handler.on_connection_event(ConnectionEvent::LocalProtocolsChange(change));
-                }
-
-                *supported_protocols = new_protocols;
-
-                continue; // Go back to the top, handler can potentially make progress again.
+            let current_proocols = supported_protocols.len();
+            supported_protocols.extend(handler.listen_protocol().upgrade().protocol_info());
```

Comment:
I'd be curious whether or not it makes a big difference if we collect `handler.listen_protocol().upgrade().protocol_info()` into a `HashSet` instead so that we can continue to use `.difference`.

We can make a permanent `HashSet` to reuse the allocation. That should be equivalent to `extend`ing a `Vec`.

Reply:
Considering we probably don't store more then few protocols, `HashSet` seems unnecessary, O(1) is better then O(n) in theory but in practice the `HashSet` has O(1 * q) where q can exceed n if n is small, which it is.

Reply:
Actually, now that I looked at `SelectUpgrade` you do have a point, composing protocols scales the n, I can probably change it to HashSet, actually it will most likely be even better since I noticed I can use the lazy nature of `difference`